### ISSUE TITLE
Encryption, compression, etc.

### ISSUE THREAD

**Issue by CrazyHoodlem** [OsamaAbbas/bytenode#99]
Hello. 

I use `bytenode` for my project and through my usage, I have discovered one feature I would like looked into. I am not too happy how the script's `cachedData` contains intact function names, variable names, and strings. I would just like a way to hide that information. It's not like I store passwords in my code, but the function names and strings do expose how my program works. 

I tried modifying this module with `zlib` to hide the raw data, but I encountered one error after another. That is why I am here.

An example of an exposed function:
```
... Qfz»ê5   enableRemoteModuleGQc* ...
```

Some of the errors I confronted that I couldn't solve (believe me, I tried):
```
ReferenceError: Require is not defined
```
```
Error: Invalid or incompatible cached data
```
Please note: after removing the `zlib` code everything works as expected.

If you are wondering why I don't just use an obfuscator as the "middleware," it is because the obfuscators I tried did not obfuscate `regex`. That was a big problem for me.

All help is appreciated.

**Comment by OsamaAbbas** [comment]
Can you please provide a minimal reproducible example that shows what you are trying to achieve and what is the problem?

Without knowing how do you use the `zlib` nor how do you "obfuscate" `regex`, I believe there is not much that I can help with.

**Comment by CrazyHoodlem** [comment]
I am not at my main computer right now (I am in a class), but I believe I
still recall what I did:

This is the first thing I tried:
```js
const fs = require("fs");
const zlib = require("zlib");
const bytenode = require("./bytenode"); // for this example
require("bytenode") would work too, as I had not yet modified anything in
the source code
const compileFile = (filePath, outputPath) => {
    return fs.writeFileSync(
        filePath,
        zlib.deflateSync(
            bytenode.compileCode(fs.readFileSync(filePath, "utf-8"))
        )
    );
};
const runFile = (filePath) => {
    return bytenode.runByteCode(
        zlib.inflateSync(
            fs.readFileSync(filePath)
        )
    );
};

// Works as expected
compileFile("any-file.js", "any-file.min.jsc");
// Throws an error; `require` is not defined I believe.
runFile("any-file.min.jsc");
```

> nor how do you "obfuscate" regex

Yeah, I don't think you can, which is why I am not using an obfuscator.

**Comment by CrazyHoodlem** [comment]
I am not at my main computer right now (I am in a class), but I believe I still recall what I did:

This is the first thing I tried:

```
const fs = require("fs");
const zlib = require("zlib");
const bytenode = require("./bytenode"); // for this example require("bytenode") would work too, as I had not yet modified anything in the source code
const compileFile = (filePath, outputPath) => {
    return fs.writeFileSync(
        filePath,
        zlib.deflateSync(
            bytenode.compileCode(fs.readFileSync(filePath, "utf-8"))
        )
    );
};
const runFile = (filePath) => {
    return bytenode.runByteCode(
        zlib.inflateSync(
            fs.readFileSync(filePath)
        )
    );
};

// Works as expected
compileFile("any-file.js", "any-file.min.jsc");
// Throws an error; 'require' is not defined I believe.
runFile("any-file.min.jsc");
```

> nor how do you "obfuscate" regex

Yeah, I don't think you can, which is why I am not using an obfuscator.

**Comment by OsamaAbbas** [comment]
I will look into it.

I like the idea too. =)

**Comment by OsamaAbbas** [comment]
Your code has some minor issues, so after fixing them it should go like this:

```javascript
const fs = require('fs');
const zlib = require('zlib');
const bytenode = require('bytenode');

const compileFile = (filePath, outputPath) => {
  return fs.writeFileSync(
    outputPath,
    zlib.deflateSync(
      bytenode.compileCode(fs.readFileSync(filePath, 'utf8'))
    )
  );
};

const runFile = (filePath) => {
  return bytenode.runBytecode(
    zlib.inflateSync(
      fs.readFileSync(filePath)
    )
  );
};

compileFile('./script.js', './script.jsz');

runFile('./script.jsz');
```

Now, that `script.js` was compiled using `bytenode.compileCode()` function, which does NOT produce a `cachedData` that understand the whole `require` and Node module system. That's why you get `'require' is not defined`.

If your code contains simple statements, like `console.log('whatever');`, the previous snippet will work. But if you want the `require` functionality, you should look into how we added the `.jsc` extension to the node module system, here: https://github.com/OsamaAbbas/bytenode/blob/734ec0c7788992c44689cb7161b3dca1071b673e/index.js#L173

So, instead of reading the file, you would need to do your inflating thing before.

Start by registering another extension, `.jsz` for example, like this:

```javascript
Module._extensions['.jsz'] = function (module, filename) {

  let bytecodeBuffer = zlib.inflateSync(
    fs.readFileSync(filename)
  );

  fixBytecode(bytecodeBuffer);
.
.
.
/// the rest will be the same..
```

Then, you would run your new `.jsz` files using `require('./script.jsz');`.

I hope this helps. Let me know if you have further questions.

**Comment by CrazyHoodlem** [comment]
Thanks, I had tried doing something like that earlier (but modifying the `.jsc` one, not creating a new one) and it had lots of errors. Your response made me try again, I realized I had an inconsistency in my code where I was inflating non-deflated files. Once I had that fixed, I still got a `require` is not defined error. Looking at your code more, I fixed the problem by changing

```js
bytenode.compileCode(fs.readFileSync(filePath, 'utf8'))
```

to

```js
bytenode.compileCode(
    Module.wrap(
        fs.readFileSync(filePath, 'utf8')).replace(/^#!.*/, "")
    )
);
```

Thank you, I have no more errors and my code works as expected.
